stress: failed test in cockroach/sql/sql.test: TestSchemaChangeReverseMutations

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/d9039b883cc2c9190448216f57ccfe4676d4e7e1

Stress build found a failed test:

```
=== RUN   TestSchemaChangeReverseMutations
I160614 07:03:35.386330 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160614 07:03:35.392204 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
W160614 07:03:35.399448 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160614 07:03:35.401553 server/node.go:401  store store=0:0 ([]=) not bootstrapped
I160614 07:03:35.403641 storage/replica_command.go:1531  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m39.903268959s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.403551912 +0000 UTC]
I160614 07:03:35.404208 server/node.go:327  **** cluster {091f4628-9a01-44ba-93cf-1b3a5dceefbc} has been created
I160614 07:03:35.404232 server/node.go:328  **** add additional nodes by specifying --join=127.0.0.1:34078
W160614 07:03:35.404243 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160614 07:03:35.404844 server/node.go:414  initialized store store=1:1 ([]=): {Capacity:10365558784 Available:7930916864 RangeCount:0}
I160614 07:03:35.404862 server/node.go:302  node ID 1 initialized
I160614 07:03:35.404914 storage/stores.go:287  read 0 node addresses from persistent storage
I160614 07:03:35.404973 server/node.go:535  connecting to gossip network to verify cluster ID...
I160614 07:03:35.405091 server/node.go:556  node connected via gossip and verified as part of cluster {"091f4628-9a01-44ba-93cf-1b3a5dceefbc"}
I160614 07:03:35.405119 server/node.go:355  [node=1] Started node with [[]=] engine(s) and attributes []
I160614 07:03:35.405142 server/server.go:417  starting https server at 127.0.0.1:45886
I160614 07:03:35.405152 server/server.go:418  starting grpc/postgres server at 127.0.0.1:34078
I160614 07:03:35.411670 storage/split_queue.go:96  splitting range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I160614 07:03:35.427403 server/updates.go:156  No previous updates check time.
I160614 07:03:35.429802 storage/replica_command.go:2079  initiating a split of range=1 [/Min-/Max) at key /Table/11
I160614 07:03:35.440524 storage/replica_command.go:1531  range 2: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m40.004532059s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.440481827 +0000 UTC]
I160614 07:03:35.440670 storage/replica_command.go:2079  initiating a split of range=2 [/Table/11-/Max) at key /Table/12
I160614 07:03:35.446229 storage/replica_command.go:1531  range 3: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m40.004532059s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.446177792 +0000 UTC]
I160614 07:03:35.446755 storage/replica_command.go:2079  initiating a split of range=3 [/Table/12-/Max) at key /Table/13
I160614 07:03:35.453463 storage/replica_command.go:1531  range 4: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m40.004532059s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.453424559 +0000 UTC]
I160614 07:03:35.453573 storage/replica_command.go:2079  initiating a split of range=4 [/Table/13-/Max) at key /Table/14
I160614 07:03:35.520372 storage/replica_command.go:1531  range 5: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m40.070008141s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.520333781 +0000 UTC]
I160614 07:03:35.523304 storage/split_queue.go:96  splitting range=5 [/Table/14-/Max) at keys [/Table/50/0]
I160614 07:03:35.523688 storage/replica_command.go:2079  initiating a split of range=5 [/Table/14-/Max) at key /Table/50
I160614 07:03:35.549076 storage/replica_command.go:1531  range 6: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m40.098101205s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.548827945 +0000 UTC]
I160614 07:03:35.549433 storage/split_queue.go:96  splitting range=6 [/Table/50-/Max) at keys [/Table/51/0]
I160614 07:03:35.550864 storage/replica_command.go:2079  initiating a split of range=6 [/Table/50-/Max) at key /Table/51
I160614 07:03:35.593264 sql/lease.go:250  publish (count leases): descID=51 name=test version=1 count=1
I160614 07:03:35.649197 sql/lease.go:250  publish (count leases): descID=51 name=test version=2 count=1
I160614 07:03:35.673372 storage/replica_command.go:1531  range 7: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407191h3m40.223085132s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 07:03:35.673327761 +0000 UTC]
W160614 07:03:35.674085 sql/schema_changer.go:293  reversing schema change due to irrecoverable error: null value in column "a" violates not-null constraint
W160614 07:03:35.675755 sql/schema_changer.go:477  reverse schema change mutation: {Descriptor_:0xc82016d458 State:WRITE_ONLY Direction:ADD MutationID:1}
W160614 07:03:35.675792 sql/schema_changer.go:477  reverse schema change mutation: {Descriptor_:0xc82016d460 State:WRITE_ONLY Direction:ADD MutationID:1}
W160614 07:03:35.675837 sql/schema_changer.go:523  delete schema change mutation: {Descriptor_:0xc82016d468 State:DELETE_ONLY Direction:ADD MutationID:2}
W160614 07:03:35.675869 sql/schema_changer.go:523  delete schema change mutation: {Descriptor_:0xc82016d478 State:DELETE_ONLY Direction:ADD MutationID:3}
I160614 07:03:35.687285 sql/lease.go:250  publish (count leases): descID=51 name=test version=4 count=1
I160614 07:03:35.706062 sql/lease.go:250  publish (count leases): descID=51 name=test version=4 count=1
I160614 07:03:35.772100 sql/lease.go:250  publish (count leases): descID=51 name=test version=5 count=1
W160614 07:03:35.794364 sql/schema_changer.go:745  Error executing schema change: null value in column "a" violates not-null constraint
I160614 07:03:35.848799 sql/lease.go:250  publish (count leases): descID=51 name=test version=6 count=1
I160614 07:03:35.887963 sql/lease.go:250  publish (count leases): descID=51 name=test version=7 count=1
I160614 07:03:35.936662 sql/lease.go:250  publish (count leases): descID=51 name=test version=8 count=1
I160614 07:03:35.975320 sql/lease.go:250  publish (count leases): descID=51 name=test version=9 count=1
W160614 07:03:35.985450 sql/lease.go:1111  error purging leases for table 51(test): server/node.go:804: node 1 stopped
W160614 07:03:35.994513 sql/schema_changer.go:274  node unavailable; try another peer
W160614 07:03:35.994617 sql/schema_changer.go:197  node unavailable; try another peer
--- FAIL: TestSchemaChangeReverseMutations (0.63s)
	schema_changer_test.go:984: incorrect columns: [k]
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
0 runs so far, 0 failures, over 30s
0 runs so far, 0 failures, over 35s
0 runs so far, 0 failures, over 40s
0 runs so far, 0 failures, over 45s
10 runs so far, 0 failures, over 50s
15 runs so far, 0 failures, over 55s
16 runs so far, 0 failures, over 1m0s
16 runs so far, 0 failures, over 1m5s
16 runs so far, 0 failures, over 1m10s
16 runs so far, 0 failures, over 1m15s
16 runs so far, 0 failures, over 1m20s
16 runs so far, 0 failures, over 1m25s
16 runs so far, 0 failures, over 1m30s
18 runs so far, 0 failures, over 1m35s
27 runs so far, 0 failures, over 1m40s
31 runs so far, 0 failures, over 1m45s
32 runs so far, 0 failures, over 1m50s
32 runs so far, 0 failures, over 1m55s
32 runs so far, 0 failures, over 2m0s
32 runs so far, 0 failures, over 2m5s
32 runs so far, 0 failures, over 2m10s
32 runs so far, 0 failures, over 2m15s
32 runs so far, 0 failures, over 2m20s
36 runs so far, 0 failures, over 2m25s
46 runs so far, 0 failures, over 2m30s
48 runs so far, 0 failures, over 2m35s
48 runs so far, 0 failures, over 2m40s
48 runs so far, 0 failures, over 2m45s
48 runs so far, 0 failures, over 2m50s
48 runs so far, 0 failures, over 2m55s
48 runs so far, 0 failures, over 3m0s
48 runs so far, 0 failures, over 3m5s
50 runs so far, 0 failures, over 3m10s
56 runs so far, 0 failures, over 3m15s
62 runs so far, 0 failures, over 3m20s
64 runs so far, 0 failures, over 3m25s
64 runs so far, 0 failures, over 3m30s
64 runs so far, 0 failures, over 3m35s
64 runs so far, 0 failures, over 3m40s
64 runs so far, 0 failures, over 3m45s
64 runs so far, 0 failures, over 3m50s
64 runs so far, 0 failures, over 3m55s
66 runs so far, 0 failures, over 4m0s
75 runs so far, 0 failures, over 4m5s
79 runs so far, 0 failures, over 4m10s
80 runs so far, 0 failures, over 4m15s
80 runs so far, 0 failures, over 4m20s
80 runs so far, 0 failures, over 4m25s
80 runs so far, 0 failures, over 4m30s
80 runs so far, 0 failures, over 4m35s
80 runs so far, 0 failures, over 4m40s
80 runs so far, 0 failures, over 4m45s
83 runs so far, 0 failures, over 4m50s
92 runs so far, 0 failures, over 4m55s
96 runs so far, 0 failures, over 5m0s
96 runs so far, 0 failures, over 5m5s
96 runs so far, 0 failures, over 5m10s
96 runs so far, 0 failures, over 5m15s
96 runs so far, 0 failures, over 5m20s
96 runs so far, 0 failures, over 5m25s
96 runs so far, 0 failures, over 5m30s
97 runs so far, 0 failures, over 5m35s
100 runs so far, 0 failures, over 5m40s
104 runs completed, 1 failures, over 5m43s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/sql/schema_changer_test.go
+++ b/sql/schema_changer_test.go
@@ -961,30 +961,39 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 	atomic.StoreUint32(&enableAsyncSchemaChanges, 1)
 
 	// Wait until all the mutations have been processed.
+	var rows *gosql.Rows
+	expectedCols := []string{"k", "b"}
 	util.SucceedsSoon(t, func() error {
 		// Read table descriptor.
 		tableDesc = sqlbase.GetTableDescriptor(kvDB, "t", "test")
 		if len(tableDesc.Mutations) > 0 {
 			return util.Errorf("%d mutations remaining", len(tableDesc.Mutations))
 		}
+
+		// Verify that t.test has the expected data. Read the table data while
+		// ensuring that the correct table lease is in use.
+		var err error
+		rows, err = sqlDB.Query(`SELECT * from t.test`)
+		if err != nil {
+			t.Fatal(err)
+		}
+		cols, err := rows.Columns()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Ensure that sql is using the correct table lease.
+		if len(cols) != len(expectedCols) {
+			return util.Errorf("incorrect columns: %v", cols)
+		}
+		if cols[0] != expectedCols[0] || cols[1] != expectedCols[1] {
+			t.Fatalf("incorrect columns: %v", cols)
+		}
 		return nil
 	})
 
-	// Verify that t.test has the expected data.
-	rows, err := sqlDB.Query(`SELECT * from t.test`)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	cols, err := rows.Columns()
-	if err != nil {
-		t.Fatal(err)
-	}
-	if len(cols) != 2 || cols[0] != "k" || cols[1] != "b" {
-		t.Fatalf("incorrect columns: %v", cols)
-	}
-
-	vals := make([]interface{}, len(cols))
+	// rows contains the data; verify that it's the right data.
+	vals := make([]interface{}, len(expectedCols))
 	for i := range vals {
 		vals[i] = new(interface{})
 	}
@@ -1025,7 +1034,7 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 
 	// Check that the index on b eventually goes live even though a schema
 	// change in front of it in the queue got purged.
-	rows, err = sqlDB.Query(`SELECT * from t.test@test_b_key`)
+	rows, err := sqlDB.Query(`SELECT * from t.test@test_b_key`)
 	if err != nil {
 		t.Fatal(err)
 	}

